Analog rewrite

arfrie22 opened the pull request:
## Summary

This PR prepares firmware 2.2.1(5), a major update focused on support of the Sensor Expander and the Expanded Sensors API, while also improving scheduling, OSPi/Linux support, networking reliability, build tooling, and documentation.

  ## Expanded Sensors

  - Add a unified Expanded Sensor framework with stable UUID-based references.
  - Add ADS1115 analog sensors, including linear, piecewise-linear, and sensor-specific subtypes.
  - Add Aggregate, System Internal, and Onboard Digital sensor types.
  - Add sensor status and validity tracking, unit conversion, configurable logging, and program watering adjustments.
  - Add sensor metadata and API endpoints for configuration, readings, adjustment previews, and log access.
  - Add paginated sensor-log retrieval and deletion for responsive operation with large datasets and OTC connections.

  ## Controller and Scheduling

  - Add SN3 and SN4 support on compatible hardware.
  - Extend master support from two to four masters.
  - Allow effective runtimes after weather or sensor adjustment to exceed 16-bit and 18-hour limits, up to the seven-day runtime cap.
  - Correct repeated Run-Once program attribution in logs and last-run data.
  - Fix parallel-group scheduling bounds and station-index validation.

  ## OSPi/Linux and Networking

  - Replace platform-dependent time and duration types with fixed-width types, fixing logging on 32-bit Linux systems with 64-bit `time_t`.
  - Update and pin OTF 0.2.2 and the maintained TinyWebsockets fork.
  - Fix Linux WebSocket descriptor leaks and improve OTC/network recovery.
  - Restore OSPi GPIO initialization and correct ADS1115 I2C detection.
  - Build container images as OSPI rather than DEMO so GPIO support is present.
  - Improve firmware update scripts and deterministic submodule handling.

  ## Build, Security, and Maintenance

  - Fix memory/resource leaks, parsing bounds, timer rollover handling, and several API validation issues.
  - Refactor notifier, sensor, built-in sensor, and option definitions to reduce duplication and simplify future expansion.

  ## Documentation

  - Add the firmware 2.2.1(5) User Manual and API Reference.
  - Add dedicated Zone Expander and Sensor Expander manuals.
  - Add stable User Manual and API links.
  - Reorganize navigation and common firmware update/compilation documentation.
  - Document Expanded Sensors, four masters, SN3/SN4, longer effective runtimes, pagination, and the updated UI workflows.

  ## Issues Resolved

  Closes #417
  Closes #415
  Closes #403
  Closes #397
  Closes #418

  ## Superseded Pull Requests

  The following pull requests are superseded or incorporated by this PR and should be closed manually after merge:

  - #409, #406, and #383: Linux `time_t` and log-format fixes are covered by the fixed-width time-type changes.
  - #370: test-mode credential handling is implemented using an ignored local header and tracked example file.
  - #324: applicable compiler-warning fixes have been incorporated or made obsolete by the updated code.
  - #323: stale project-reorganization proposal; the exact directory layout proposed there was not adopted.

arfrie22 committed:
start scaffolding for sensor

arfrie22 committed:
ads1115 support for esp and ospi

arfrie22 committed:
Wrote code for esp from scratch

arfrie22 committed:
fix submodules

arfrie22 committed:
fix i2cd.h

arfrie22 committed:
fix print

arfrie22 committed:
fix import

arfrie22 committed:
added missing semicolon

arfrie22 committed:
add virtual deconsturctor for analogsensor interface

arfrie22 committed:
only include i2c in the c block for i2cd.h

arfrie22 committed:
fix interface again

arfrie22 committed:
add mutex for i2cd.h

arfrie22 committed:
fix init

arfrie22 committed:
fix i2c

arfrie22 committed:
Working ads 1115 impl

arfrie22 committed:
test smt50

arfrie22 committed:
add sensor class

arfrie22 committed:
inital setup of sensors and endpoint to get them

arfrie22 committed:
add delete sensor

arfrie22 committed:
move sensors over to define to allow them to be disabled

arfrie22 committed:
current analog setup with endpoints

arfrie22 committed:
added logs and saving / loading of sensors from flash

arfrie22 committed:
Store sensor data in flash not memory and make weather sensor class log return newest first

arfrie22 committed:
remove serial.printf

arfrie22 committed:
change pagination to cursor instead of skip for reading sensor log

arfrie22 committed:
Current sensor code, littlefs makes this too slow on each read/write

arfrie22 committed:
Temp disabled logging sensors since it would add 4s of latency

arfrie22 committed:
split log file into multiple files to make it a reasonable speed on littlefs

arfrie22 committed:
added HTML_INTERNAL_ERROR (0x50) for error codes

arfrie22 committed:
switch include bracket to quote

arfrie22 committed:
add cmath to fix nan issue on docker

arfrie22 committed:
fix printf type from sizet to uint16t

arfrie22 committed:
add file commands for linux

arfrie22 committed:
fix read write to not cause race condition

arfrie22 committed:
fix fwrite

arfrie22 committed:
add i2c detect

arfrie22 committed:
add limits include for the server

arfrie22 committed:
fix demo using bus

arfrie22 committed:
Fix sensor reading for more than 1 sensor, made it so sensor's inital values dont have to be 0 (useful for min on an ensemble sensor) made ensemble sensors have 4 inputs that can be scaled

arfrie22 committed:
Added sensor adjustments for programs

arfrie22 committed:
gate the adjustments inside of a defined block for demo

arfrie22 committed:
removed dsa which doesn't exist

arfrie22 committed:
allowed jsn to give information specific to sensor type

arfrie22 committed:
fix json

arfrie22 committed:
Try to fix compilation on docker

arfrie22 committed:
update otf

arfrie22 committed:
update otf

arfrie22 committed:
move numeric limits to above bfiller.h

arfrie22 committed:
include types.h

arfrie22 committed:
added missing include for opensprinkler server

arfrie22 committed:
added missing stdarg

arfrie22 committed:
switch to using points to describe function instead of piecewise

arfrie22 committed:
ensure that points are entered in order

arfrie22 committed:
use std limits infinity

arfrie22 committed:
added sensor type to json

arfrie22 committed:
add sensors flags to disable sensor/logging, and fixed the json output from ensemble sensor children

arfrie22 committed:
add clearing sensor log

arfrie22 committed:
make it so sensor adjustments can't have two points with the same x value

arfrie22 committed:
start of dynamically generating sensor page

arfrie22 committed:
faught with C++ to make it easier to print out enum names

arfrie22 committed:
update enums

arfrie22 committed:
Update sensor description json

arfrie22 committed:
fix json

arfrie22 committed:
add default option to json

arfrie22 committed:
flush buffer to prevent stack smashing

arfrie22 committed:
fix ensemble sensor

arfrie22 committed:
add default option to flags

arfrie22 committed:
fixed ads1115 pin json

arfrie22 committed:
made ensemble sensor use a drop down to select sensors

arfrie22 committed:
made sensor interval in minutes, fixed reading sensors, made ensemble sensor support -1 for a child to refer to itself, fixed flags json added jsd endpoint

arfrie22 committed:
fix null ptr and move sen adj to be in the program endpoint

arfrie22 committed:
Merge remote-tracking branch 'origin/dev/221_3' into analog-rewrite

rayshobby committed:
merge 2.2.1(4)

rayshobby committed:
remove AVR support

rayshobby committed:
remove AVR from this firmware

rayshobby committed:
fix merge conflicts

rayshobby committed:
fix dec2hexchar issue

rayshobby committed:
fix macro defines so the firmware compiled in demo

rayshobby committed:
fix indentation and fix typo (inital->initial)

rayshobby committed:
fix minor issues with i2cd

rayshobby committed:
fix ospi compilation error

rayshobby committed:
make time type consistent (uint32_t)

rayshobby committed:
recover sensor.cpp back to original for saving RAM/flash memory consumption

rayshobby committed:
use stack variables

rayshobby committed:
change sensor value type from double to float; set default sampling interval to 5 minutes; default max to 5000; refactor sensor files initialization to account for firmware update within dotted version; change sensor_adjustment allocation to static to avoid repeatedly calling new and delete

rayshobby committed:
adopt placement new to allow statically allocated sensor object, eliminating dynamic new and delete every time we need to read a sensor

rayshobby committed:
change notifier to use circular array queue instead of linked list, to avoid new and delete repeatedly

rayshobby committed:
fix demo compilation error

rayshobby committed:
fix docker build error

rayshobby committed:
adjust the conversative size of tmp_buffer and ether_buffer

rayshobby committed:
adjust overflow buffer size; improve email feature to support local smtp server that does not require authentication

rayshobby committed:
revise sensor log, change log data structure, optimize each log file to be no more than one littleFS block (8K) in size

rayshobby committed:
move sensor logs to /logs/ subfolder; set a cap to the sprinkler log size (to max 1.2MB) to make room for sensor logs; add comments to clarify the new sensor class and variables are for analog sensors (distinguish from onboard sensors)

rayshobby committed:
add support for Master3 and Master4; add support for expanding iopt options without having to trigger a factory reset

rayshobby committed:
restructure sensor adjustment data file; change to lazy file creation; move relevant operations from opensprinkler_server.cpp to program.h and cpp as program data and sensor adjustment are supposed to be fully in sync

rayshobby committed:
add uuid for sensors to reference them by unique ids instead of positional indices. this makes program adjustments and ensemble sensors more robust to sensor changes such as deletions

rayshobby committed:
reorganize SensorAdjustment class, move its read and write functions to sensor.cpp, use direct binary buffer read/write as (unlike Sensor classes) it doesn't use polymorphism

rayshobby committed:
rearrange some sensor related functions to the sensor.h and .cpp files for better organization; change sens.dat to use lazy writing (grows as necessary); improve sensor APIs to support modification and deletion using either positional index or uuid

rayshobby committed:
fix sens.dat writing bug; add utility API to list and delete flash file on ESP8266

rayshobby committed:
fix bug in get_sensor_adjustment; fix where get_sensor_adjustment is called

rayshobby committed:
add rate limit to poll_sensors in main loop; add sensor_adjustment factor to notification channel

rayshobby committed:
fix sensor add bug

rayshobby committed:
fix sensor find_index bug; reformat sensor adj in /jp output

rayshobby committed:
update sensor log endpoint names to make them consistent with the other naming conventions

rayshobby committed:
change bfill to use res.write callback, this eliminates the need to hold the output content in ether_buffer first, instead, it directly outputs to otf buffer, which already handles content streaming. as a result, there is no need to check available_etherbuffer_size, and we can use smaller ether_buffer in the future

rayshobby committed:
support multiple formats of /jsl output, in particular, it support a binary dump which the UI can utilize to minimize log data size

rayshobby committed:
refactor sensors: put them in sensors subfolder. update sensor descriptions, streamline sensor unit, group, short name by replacing long switch-case sequence with macros;

rayshobby committed:
add missing sensors folder

rayshobby committed:
rename some variable and class names; remove scale and offset from base class and move them into specific classes like ADS1115 as not all sensors need scale and offset (AggregateSensor and WeatherSensor generally don't need that)

rayshobby committed:
remove min/max from aggregate sensor children fields as those are not necessary

rayshobby committed:
modify sens.dat layout to make it more future proof (in case we need to expand the base class members and subclass members

rayshobby committed:
add sensor::status, supporting indicator of whether sensor reading is valid/error/stable; remove get_initial_value now that sensor status is in place; add /jpa endpoint to send program adjustments (including weather-based and sensor-based) to UI to facilitate preview

rayshobby committed:
add 'usa' parameter for manual_start_program to allow applying sensor adjustment on manually started programs; update sensor::flag to uint8_t to be consistent with sensor_memory_t::flag; improve write_buf and read_buf to eliminate the need of explicitly specifying type

rayshobby committed:
remove redundant check

rayshobby committed:
rename SensorAdjustment::flag to reserved since for now this variable is not used (enabled can be implied from the uuid value)

rayshobby committed:
revert back to use SensorAdjustment::flag as it allows indepenently turning off adjustment without losing parameters; update code to allow duplicate x value in case the user wants a sharp threshold operation

rayshobby committed:
suport mock ads1115 for LInux/Demo; given all platforms now support sensors, remove USE_SENSORS and USE_ADS1115 macros; fix millis issue on 64-bit linux;

rayshobby committed:
temporarily change the forward declaration of millis back until we changes to OTF are committed upstream

rayshobby committed:
fix missing sensor load for linux; add sensor description hints for UI

rayshobby committed:
expand ADS1115 sensor by adding subtypes (specific sensors with baked in formula such as SMT50, as well as general formula like piecewise linear); various optimization to reduce /jsd output size; add System Internal Sensor (i.e. free heap, free flash, CPU temperature); add support for temperature conversion within the temperature group

rayshobby committed:
add support for OnBoardDigitalSensor (SN1, SN2 etc); add Heap Fragmentation to supported SystemInternalSensor

rayshobby committed:
fix potential memory leak issues in EmailSender

rayshobby committed:
refactor Notifier class to make it easy to read and maintain for the future; stagger sensor read at reboot to avoid all sensors polling at the same time; fix RSSI sensor so its reading is realiable

rayshobby committed:
Bump OTF submodule to develop-0.2.1 (WIP for 0.2.1 release)

rayshobby committed:
add supports for sn3 and sn4, refactor iopts table to consolidate all of name, prompt, max into a central place making it easy to add options in the future

rayshobby committed:
add SN3 and SN4 support, consolidate SNi sensor code paths to reduce code duplication; add 'hwd' (hardware detected' flag for ADS1115

rayshobby committed:
update all non size specific types (long, unsigned long) to size specific ones

rayshobby committed:
fix OSPi branch of ads1115 which results in i2c detection result being flipped

rayshobby committed:
fix a bug where the sensor continues to log even when the sensor is invalid

rayshobby committed:
support adjusted runtimes beyond 16-bit limit

Store effective queue and log durations as uint32_t while keeping programmed durations unchanged for persistence compatibility. Apply weather and sensor scaling with a bounded helper so adjusted runtimes cannot overflow.

Keep manual and remote commands capped at 18 hours through a shared constant, emit full-width /jc duration values, and document the remote auto-refresh requirement identified during PR review.

rayshobby committed:
restore OSPi station GPIO initialization

Restore shift-register GPIO claims and the latch rising edge that were inadvertently removed with AVR support. Preserve the alternate data pin used by early Raspberry Pi revisions.

salbahra commented on opensprinkler_server.cpp:
```diff
@@ -1173,41 +1062,61 @@ void server_json_programs_main(OTF_PARAMS_DEF) {
+		float wa = get_program_water_percent(prog) / 100.f;
+		float sa = get_program_sensor_adj(pid);
+		if(pid) bfill.emit_p(PSTR(","));
+		bfill.emit_p(PSTR("{\"wa\":$E,\"sa\":$E,\"ta\":$E}"), wa, sa, wa * sa);
```
**Expose the effective-runtime limit alongside `/jpa` factors**

`water_time_scale()` now caps each effective station runtime at `MAX_RUNTIME_DURATION` (604,800 seconds), but `/jpa` exposes only the uncapped `wa`, `sa`, and `ta` factors. [App PR #294 consumes these factors in its scheduled-program preview](https://github.com/OpenSprinkler/OpenSprinkler-App/pull/294#discussion_r3609177544) and currently computes `base x wa x sa` without the firmware cap, so it can display a longer run than the controller will execute. For example, 64,800 x 2.5 x 4 is 648,000 seconds in the preview versus 604,800 seconds in the queue. Since the limit is per station and cannot be represented by simply clamping `ta`, could `/jpa` expose the maximum effective runtime, such as a top-level `max_runtime`, or otherwise provide a stable API contract that the App can apply as `min(computed, max_runtime)`? This keeps preview and execution aligned without hard-coding a firmware-only constant in the App.

rayshobby replied:
This has been addressed in the latest code.

rayshobby committed:
expose runtime cap in program adjustment API

Add maxrt to /jpa so UI previews can apply the same effective station runtime limit as the firmware. Document the response field and preview calculation.

rayshobby committed:
fix repeated run-once program attribution

salbahra commented on main.cpp:
```diff
@@ -874,47 +782,39 @@ void do_loop()
 				unsigned char runcount = prog.check_match(curr_time, &will_delete);
 				if(runcount>0) {
+					const bool repeated_runonce =
+						strncmp_P(prog.name, PSTR(RUNONCE_REPEAT_PREFIX), sizeof(RUNONCE_REPEAT_PREFIX) - 1) == 0;
```
**Do not classify run-once programs by a user-controlled name**

`/cp` accepts arbitrary program names, so an ordinary scheduled program whose name begins with `Run-Once with repeat` is classified here as run-once and receives `RUNONCE_PID`. `process_dynamic_events()` skips every queue PID at or above `MANUAL_PID`, which means that normal program can bypass controller-disabled, rain-delay, and per-sensor shutdown handling in addition to being misattributed in logs and notifications. Please identify generated repeated run-once programs through explicit internal metadata rather than a name prefix, and add a regression using an ordinary program with this prefix.

rayshobby replied:
Programs created by users are not named like "Run-once with repeat". This is purely auto-created run-once with repeat programs.

salbahra commented on opensprinkler_server.cpp:
```diff
@@ -1809,12 +1644,12 @@ void server_change_manual(OTF_PARAMS_DEF) {
 		if (findKeyVal(FKV_SOURCE, tmp_buffer, TMP_BUFFER_SIZE, PSTR("t"), true)) {
-			timer=(uint16_t)atol(tmp_buffer);
-			if (timer==0 || timer>64800) {
+			timer=strtoul(tmp_buffer, NULL, 10);
```
**Validate manual durations before narrowing to `uint32_t`**

On 64-bit Linux and OSPi, `strtoul("4294967297")` produces a 64-bit value that is narrowed to `1` in `timer` before the 18-hour bound is checked, so malformed oversized input is accepted. The same pattern exists in `mqtt.cpp` at the manual-run timer parse. Please parse into a wide type, validate the end pointer and range before casting, ideally through one shared helper, and regress both HTTP and MQTT inputs above `UINT32_MAX`.

salbahra commented on opensprinkler_server.cpp:
```diff
@@ -2077,78 +1880,865 @@ void server_pause_queue(OTF_PARAMS_DEF) {
+	uint32_t flat_idx = 0;
+	uint32_t count = 0;
+
+	for (uint16_t fi = 0; fi < total_files && count < max_count; fi++) {
```
**Avoid scanning all retained history for an `after` query**

[App PR #294 now uses a one-day `after` cutoff for its initial chart load](https://github.com/OpenSprinkler/OpenSprinkler-App/pull/294#discussion_r3609177224), but this loop starts with the oldest file and only rejects pre-window records after reading each one. Linux and OSPi can retain 50 x 65,535 = 3,276,750 records, so even a one-day refresh may read nearly the entire log before returning the bounded response. Please use per-file timestamp bounds or newest-first traversal until the cutoff, while preserving response order, and add a behavioral check that a recent window does not read every retained file.

rayshobby replied:
Addressed in 80aa7da. Queries using after now inspect each file’s newest record and skip the entire file when it predates the requested window, preserving chronological output without scanning all retained records. Also, Linux/OSPi retention was reduced to 50 × 16,384 records.

salbahra commented on mqtt.cpp:
```diff
@@ -206,7 +202,7 @@ void manualRun(char *message){
 				q->dur = timer;
 				q->sid = sid;
-				q->pid = 99;
+				q->pid = MANUAL_PID;
```
**Reject Master 3 and Master 4 in MQTT manual runs**

HTTP `/cm` rejects all four configured master stations, but this MQTT manual-run path checks only `mas` and `mas2` before reaching this enqueue. A station configured as Master 3 or Master 4 can therefore be run independently through MQTT even though the equivalent HTTP request is rejected. Please use the same four-master predicate, preferably a shared helper, and add MQTT coverage for each configured master.

rayshobby replied:
This has been addressed in the latest code.

rayshobby committed:
optimize sensor log queries and master validation

Skip sensor log files outside after windows and seek directly to deep cursors. Reduce Linux log capacity to 16,384 records per file and validate stored ring geometry. Use the shared master predicate across scheduling and manual MQTT/HTTP paths.

rayshobby committed:
add 2.2.1(5) documentation and dedicated Sensor Expander section

Freeze the 2.2.1(4) manual and API to their as-shipped state, moving the
runtime-cap, extended-duration, and /jpa content that had been added there
into new 2.2.1(5) copies. Document the new external sensor endpoints (/jsn,
/csn, /dsn, /jsl, /dsl, /jsd) in the 2.2.1(5) API reference. Add a standalone
top-level Sensor Expander section (overview, hardware, configuration, and
sensor-based watering). Reorder the nav to list 2.2.1(5) first and ignore the
generated mkdocs site/ output.

rayshobby committed:
add reliable sensor log pagination

Add opt-in physical-slot pagination to /jsl with cursor, total, completion, and CORS-exposed response headers. Preserve legacy behavior and document time-filter restrictions.

rayshobby committed:
add paginated sensor log deletion

Add opt-in cursor pagination and progress responses to UUID-based /dsl clearing. Batch record updates through the shared scratch buffer to keep controllers responsive without increasing ESP8266 RAM usage.

rayshobby committed:
support time windows in sensor log pagination

rayshobby committed:
improve LCD startup splash

rayshobby committed:
update OTF and fix Linux websocket descriptor leaks

rayshobby committed:
ensure container images build the OSPI target

rayshobby committed:
harden input parsing and runtime resource handling

rayshobby committed:
integrate OTF 0.2.2 network safeguards

rayshobby committed:
preserve timestamps in sensor log tombstones

rayshobby committed:
pin OTF submodule to 0.2.2

rayshobby committed:
keep test mode credentials local

rayshobby committed:
improve sensor fallback and controller diagnostics

Default sensor adjustments to 100 percent when readings are unavailable or invalid. Show hardware and firmware versions on the Button 2 MAC screen, disable debug output for production builds, and avoid truncated Linux log paths by deleting relative to the open directory.

rayshobby committed:
fix OSPi startup splash

Display the OSPi hardware label instead of interpreting its platform identifier as a numeric revision. Initialize the Linux SSD1306 drawing color so the startup logo renders correctly.

rayshobby committed:
harden OSPi firmware updates

Require fast-forward updates, stop on failures, and synchronize and check out the exact pinned submodule revisions before building. Apply the same deterministic submodule handling to standalone builds.

rayshobby committed:
fix station group scheduling and API validation

Restore password authentication for the change-password endpoint so a caller must supply the current password before setting a replacement.

Validate /cs station IDs with strict numeric parsing and reject sid == nstations, preventing an out-of-bounds station access. Restrict station group values to sequential groups 0-3 or the parallel group 255.

Treat only valid sequential group IDs as sequential and guard all sequential timing arrays from parallel-group indexing. This fixes out-of-bounds accesses while stopping, shifting, pausing, scheduling, or applying master timing adjustments to parallel stations.

Exclude parallel and remote-extension stations from sequential-group staggering and last-stop-time bookkeeping while retaining their concurrent one-second startup stagger.

rayshobby committed:
guard file deletion endpoint in debug builds

rayshobby committed:
update firmware 2.2.1(5) API reference

rayshobby committed:
update firmware documentation and navigation

Expand the 2.2.1(5) API and user manual for sensor, scheduling, and diagnostic changes. Consolidate the Sensor Expander guide, add a Zone Expander manual, and provide stable top-level manual and API links while retaining version-specific documentation.

rayshobby committed:
reorganize common firmware documentation

Split firmware update and compilation instructions into dedicated pages, reduce the home page to a documentation hub, and update cross-references. Add navigation badges for the new firmware and Sensor Expander documentation.

rayshobby committed:
add zone and sensor expander manuals

Document Zone Expander compatibility, wiring, indexing, and configuration. Add the Sensor Expander hardware, sensor setup, logging, status indicator, and program adjustment workflows with supporting images. Flatten the product manual paths, consolidate shared assets, and update firmware manual, API, navigation, and troubleshooting links.

rayshobby committed:
update firmware and expander documentation

rayshobby committed:
finalize firmware 2.2.1(5) user manual

rayshobby committed:
use configured aggregate sensor default action

rayshobby committed:
merge master and resolve documentation conflict

rayshobby committed:
initialize OS 3.4 sensor inputs

rayshobby committed:
harden CI and Docker release builds

Use explicit recursive checkouts and local build contexts for release images so OSPi submodules are always available. Add package permissions, attach ESP8266 firmware to releases, remove redundant build artifacts and duplicate demo compilation, and keep multi-architecture validation for non-PR builds. Exclude local, generated, runtime, and credential files from Docker contexts.

rayshobby committed:
fix PlatformIO CI dependency install

Use npm install for html-minifier-terser because master does not track package-lock.json. Pull request merge checkouts therefore cannot use npm ci.
